### ISSUE TITLE
Couldn't create audio thread (stacksize issue)

### ISSUE THREAD

**Issue by rom1v** [libsdl-org/SDL#7367]
On a real project, if I execute this piece of code, I get the error `Couldn't create audio thread`:

```c
SDL_AudioSpec desired = {
    .freq = 48000,
    .format = AUDIO_S16LSB,
    .channels = 2,
    .samples = 4096,
};
SDL_AudioSpec obtained;

SDL_AudioDeviceID device = SDL_OpenAudioDevice(NULL, 0, &desired, &obtained, 0);
if (!device) {
    fprintf(stderr, "%s\n", SDL_GetError());
}
```

I fail to reproduce on a minimal sample.

However, it is related to the stack size, because either of these patches fixes the issue:

```diff
diff --git a/src/audio/SDL_audio.c b/src/audio/SDL_audio.c
index b1b87d151..128dddac9 100644
--- a/src/audio/SDL_audio.c
+++ b/src/audio/SDL_audio.c
@@ -1542,7 +1542,7 @@ open_audio_device(const char *devname, int iscapture,
         /* Start the audio thread */
         /* !!! FIXME: we don't force the audio thread stack size here if it calls into user code, but maybe we should? */
         /* buffer queueing callback only needs a few bytes, so make the stack tiny. */
-        const size_t stacksize = is_internal_thread ? 64 * 1024 : 0;
+        const size_t stacksize = 0;
         char threadname[64];
 
         SDL_snprintf(threadname, sizeof (threadname), "SDLAudio%c%d", (iscapture) ? 'C' : 'P', (int) device->id);
```

or

```diff
diff --git a/src/thread/pthread/SDL_systhread.c b/src/thread/pthread/SDL_systhread.c
index 4a73b5a1e..af1ad731c 100644
--- a/src/thread/pthread/SDL_systhread.c
+++ b/src/thread/pthread/SDL_systhread.c
@@ -108,11 +108,6 @@ SDL_SYS_CreateThread(SDL_Thread * thread)
         return SDL_SetError("Couldn't initialize pthread attributes");
     }
     pthread_attr_setdetachstate(&type, PTHREAD_CREATE_JOINABLE);
-    
-    /* Set caller-requested stack size. Otherwise: use the system default. */
-    if (thread->stacksize) {
-        pthread_attr_setstacksize(&type, thread->stacksize);
-    }
 
     /* Create the thread and go! */
     if (pthread_create(&thread->handle, &type, RunThread, thread) != 0) {
```

On Debian sid, with kernel:
```
$ uname -r
6.1.0-5-amd64
```

A workaround is to assign a callback to `desired.callback` (this forces to use the callback behavior), because then `is_internal_thread` is `false`, so the `stacksize` is set to 0:
https://github.com/libsdl-org/SDL/blob/5b0351a672701e28c3bfb17d44a458b82e8a4175/src/audio/SDL_audio.c#L1238
https://github.com/libsdl-org/SDL/blob/5b0351a672701e28c3bfb17d44a458b82e8a4175/src/audio/SDL_audio.c#L1479

**Comment by rom1v** [comment]
I managed to reproduce on a minimal sample, by just calling `avdevice_register_all()` beforehand.

See https://github.com/rom1v/stacksize_fail

It might not be related to SDL then.

**Comment by rom1v** [comment]
In fact, it happens whenever I call any `libavdevice` function (like `avdevice_version()`). Increasing stacksize to `72 * 1024` happens to work :shrug:

**Comment by icculus** [comment]
We can bump this higher--it's an extremely conservative value--but I would like to understand what happened here too.

**Comment by weirddan455** [comment]
I'm not able to reproduce this bug on Arch Linux.  I tried with both of your samples and they are both successful.

On your 2nd smaple (where you don't call into SDL), I'm assuming it's failing for you on `pthread_create`?  If so, can you modify the code to print out the return value of the function?  It may give some information as to why it's failing.  https://man7.org/linux/man-pages/man3/pthread_create.3.html

**Comment by rom1v** [comment]
For some reason, the sample [avdevice_stackfail](https://github.com/rom1v/avdevice_stackfail) (in which I don't use SDL) now works on my machine :/

However, the one using SDL [stacksize_fail](https://github.com/rom1v/stacksize_fail) still fails, so I added a log in SDL, and `pthread_create()` returns 22:

```console
$ errno 22
EINVAL 22 Invalid argument
```

**Comment by slouken** [comment]
@icculus, did we increase the stack size for the audio thread?

**Comment by icculus** [comment]
In SDL3, it always uses zero (system-default stack size), so this should definitely be resolved if the only problem was the stack size used.

Honestly, we could probably change this for SDL2, also, and call it a day.

**Comment by slouken** [comment]
Go for it!

**Comment by icculus** [comment]
Fixed in 42d979a231527478cf316e8f6965852fece18547 for the SDL2 branch (no need to cherry pick; it isn't relevant to SDL3 and we're working towards 2.30.0 so ignoring the 2.28.x branch).
